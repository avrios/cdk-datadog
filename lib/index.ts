import { Construct } from 'constructs';
import { CfnResource } from 'aws-cdk-lib';
import * as _ from 'lodash';

/**
 * Selectively exposed based on
 * https://github.com/DataDog/datadog-cloudformation-resources/blob/datadog-monitors-monitor-4.7.1/datadog-monitors-monitor-handler/datadog-monitors-monitor.json
 */
export interface DatadogMonitorProps {
    /**
     * The type of the monitor
     */
    readonly type: 'audit alert' | 'composite' | 'event alert' | 'event-v2 alert' | 'log alert' | 'metric alert' | 'process alert' | 'query alert' |
        'service check' | 'synthetics alert' | 'trace-analytics alert' | 'slo alert' | 'rum alert' | 'ci-pipelines alert' | 'error-tracking alert' |
        'ci-tests alert';

    /**
     * The monitor query
     */
    readonly query: string;

    /**
     * Name of the monitor
     */
    readonly name?: string;

    /**
     * Tags associated with the monitor
     */
    readonly tags?: Array<string>;

    /**
     * The monitor options
     */
    readonly options?: DatadogMonitorOptions;

    /**
     * A message to include with notifications for the monitor
     */
    readonly message?: string

    /**
     * Additional parameters to pass through to the underlying CloudFormation template.
     * Allows for defining parameters that have not been exposed (yet).
     * Please note that the params will need to follow PascalCase, as they are CloudFormation identifiers that are
     * passed onward verbatim.
     */
    readonly additionalMonitorParams?: {
        [key: string]: any;
    };
}

export interface DatadogMonitorOptions {
    /**
     * The threshold definitions
     */
    readonly thresholds?: DatadogMonitorThresholds;
}

export interface DatadogMonitorThresholds {
    /**
     * Threshold value for triggering an alert
     */
    readonly critical: number;

    /**
     * Threshold value for recovering from an alert state
     */
    readonly criticalRecovery?: number;

    /**
     * Threshold value for recovering from an alert state
     */
    readonly ok?: number;

    /**
     * Threshold value for triggering a warning
     */
    readonly warning?: number;

    /**
     * Threshold value for recovering from a warning state
     */
    readonly warningRecovery?: number;
}

export class DatadogMonitor extends Construct {
    constructor(scope: Construct, id: string, props: DatadogMonitorProps) {
        super(scope, id);
        this.createMonitor(id, props);
    }

    private createMonitor(id: string, props: DatadogMonitorProps): CfnResource {
        return new CfnResource(this, id, {
            type: 'Datadog::Monitors::Monitor',
            properties:  _.merge({
                    // Note the (left-hand) PascalCase here! These are CloudFormation identifiers.
                    Type: props.type,
                    Query: props.query,
                    Name: props.name,
                    Tags: props.tags,
                    Message: props.message,
                    Options: props.options ? {
                        Thresholds: {
                            Ok: props.options?.thresholds?.ok,
                            Critical: props.options?.thresholds?.critical,
                            CriticalRecovery: props.options?.thresholds?.criticalRecovery,
                            Warning: props.options?.thresholds?.warning,
                            WarningRecovery: props.options?.thresholds?.warningRecovery,
                        }
                    } : undefined
                },
                props.additionalMonitorParams
            )
        });
    }
}
