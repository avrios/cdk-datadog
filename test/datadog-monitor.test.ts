import '@aws-cdk/assert/jest';
import { Stack } from 'aws-cdk-lib';
import * as DatadogMonitor from '../lib/index';
import { SynthUtils } from "@aws-cdk/assert";

test('Minimal Datadog Monitor Snapshot Test', () => {
    const stack = new Stack();

    // WHEN
    new DatadogMonitor.DatadogMonitor(stack, 'DatadogTestMonitor', {
        query: 'some-query',
        type: 'metric alert',
        datadogCredentials: {
            apiKey: 'some-api-key',
            applicationKey: 'some-app-key'
        }
    });

    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Datadog Monitor Can Be Extended With Unsupported Properties', () => {
    const stack = new Stack();
    const criticalThreshold = 111;
    const creatorName = 'some creator';
    const isLocked = true;

    // WHEN
    new DatadogMonitor.DatadogMonitor(stack, 'DatadogTestMonitor', {
        query: 'some-query',
        type: 'metric alert',
        options: {
            thresholds: {
                critical: criticalThreshold
            }
        },
        datadogCredentials: {
            apiKey: 'some-api-key',
            applicationKey: 'some-app-key'
        },
        additionalMonitorParams: {
            Options: {
                Locked: isLocked
            },
            Creator: {
                Name: creatorName
            }
        }
    });

    // THEN
    expect(stack).toHaveResource('Datadog::Monitors::Monitor', {
        // Nested properties can be extended with functionality not yet exposed
        Options: {
            Thresholds: {
                Critical: criticalThreshold
            },
            Locked: isLocked
        },
        // Additional properties are added
        Creator: {
            Name: creatorName
        }
    });
});
