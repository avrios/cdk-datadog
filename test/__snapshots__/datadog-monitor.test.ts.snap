// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minimal Datadog Monitor Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "DatadogTestMonitor32CCE9EB": Object {
      "Properties": Object {
        "DatadogCredentials": Object {
          "ApiKey": "some-api-key",
          "ApiURL": "https://api.datadoghq.eu",
          "ApplicationKey": "some-app-key",
        },
        "Query": "some-query",
        "Type": "metric alert",
      },
      "Type": "Datadog::Monitors::Monitor",
    },
  },
}
`;
